docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+    CONSTRAINT block_pkey PRIMARY KEY (block_header_height),
+    CONSTRAINT last_commit_height_unique UNIQUE (block_last_commit_height),
+    CONSTRAINT height_positive CHECK (block_header_height >= 0)
+) TABLESPACE pg_default;
+```
+
+Additional tables will be required to persist normalized data related to the block structure. All database schema will be available in the `rpc-companion` Github repository under the `/database/schema` folder.
+
+##### Data types
+
+This solution will implement a data schema in the database using Postgresql built-in data types. By using a relational database, there's a possibility
+to better normalize the data structures, this might provide savings in storage but might also add to the complexity of returning a particular
+dataset because the data joins that will be required. Also, it would be important ensure the referential integrity is not violated since
+this can cause issues to the clients consuming the data.
+
+In order to accurately ensure the database is storing the full node data types properly, the database might implement custom data types (`domains` in Postgresql).
+
+For example, Postgresql doesn't have an unsigned `uint64` datatype, therefore in order to support this in the database, you can use a `domain`, which is a base types with additional constrains. For example, this is the definition of a `uint64` domain:
+
+```sql
+-- DOMAIN: comet.uint64
+
+CREATE DOMAIN comet.uint64
+    AS numeric;
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_max CHECK (VALUE <= '18446744073709551615'::numeric);
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_positive CHECK (VALUE >= 0::numeric);
+```
+Additional domains for `uint32` and `uint8` will also be implemented.
+
+##### Schema migration
+
+Another point to consider is when the data structures change across CometBFT releases. There are a couple of ways to provide a solution to this
+problem.
+
+One way is to support a mechanism to migrate the old data to the new data structures. This would require additional logic for the migration process
+that would need to be run before ingesting data from an upgraded full node that contains the new data structures. But this approach might also
+cause some issues, for example, if the new data structure has a new field that there's no corresponding value in the "old" data structure, probably
+the value would need to be set to `[null]` but this can have unintended consequences.
+
+Another potential solution for this scenario is to find a way in the database that can support "versioning" of data structures. For example, let's
+assume there's a `Block` structure, let's call it `v1`. If in the future there's a need to modify this structure that is not compatible with the
+previous data structure, then the database would support a `v2` schema for `ResultBlock` and an `index` table could determine the criteria on which
+data structure should be used for inserting or querying data.
```

Comment:
I like this approach, as it is close to what we have in mind for RPC, which is what users will see.

Reply:
agree